book-content/chapters/15-designing-your-types.md

```diff
@@ -0,0 +1,1444 @@
+//                                 true/false
+```
+
+If the condition is true, it resolves to the 'true' part, just like a normal ternary. If it's false, it resolves to the 'false' part.
+
+In this case, if `T` is an array, it resolves to `T`. If it's not, it resolves to `T[]`.
+
+This means that our examples above now work as expected:
+
+```ts twoslash
+type ToArray<T> = T extends any[] ? T : T[];
+
+// ---cut---
+type Example = ToArray<string>;
+//   ^?
+
+type Example2 = ToArray<string[]>;
+//   ^?
+```
+
+Conditional types turn TypeScript's type system into a full programming language. They're incredibly powerful, but they can also be incredibly complex. You'll rarely need them in application code, but they can perform wonders in library code.
+
+## Mapped Types
+
+Mapped types in TypeScript allow you to create a new object type based on an existing type by iterating over its keys and values. This can be let you be extremely expressive when creating new object types.
+
+Consider this `Album` interface:
+
+```tsx
+interface Album {
+  name: string;
+  artist: string;
+  songs: string[];
+}
+```
+
+Imagine we want to create a new type that makes all the properties optional and nullable. If it were only optional, we could use `Partial`, but we want to end up with a type that looks like this:
+
+```tsx
+type AlbumWithNullable = {
+  name?: string | null;
+  artist?: string | null;
+  songs?: string[] | null;
+};
+```
+
+Let's start by, instead of repeating the properties, using a mapped type:
+
+```tsx
+type AlbumWithNullable = {
+  [K in keyof Album]: K;
+};
+```
+
+This will look similar to an index signature, but instead of `[k: string]`, we use `[K in keyof Album]`. This will iterate over each key in `Album`, and create a property in the object with that key. `K` is a name we've chosen: you can choose any name you like.
+
+In this case, we're then using `K` as the value of the property, too. This is not what we want eventually, but it's a good start:
+
+```tsx
+// Hovering over AlbumWithNullable shows:
+type AlbumWithNullable = {
+  name: "name";
+  artist: "artist";
+  songs: "songs";
+};
+```
+
+We can see that `K` represents the _currently iterated key_. This means we can use it to get the type of the original `Album` property using an indexed access type:
+
+```tsx
+type AlbumWithNullable = {
+  [K in keyof Album]: Album[K];
+};
+
+// Hovering over AlbumWithNullable shows:
+type AlbumWithNullable = {
+  name: string;
+  artist: string;
+  songs: string[];
+};
+```
+
+Wonderful - we've now recreated the object type of `Album`. Now we can add `| null` to each property:
+
+```tsx
+type AlbumWithNullable = {
+  [K in keyof Album]: Album[K] | null;
+};
+
+// Hovering over AlbumWithNullable shows:
+type AlbumWithNullable = {
+  name: string | null;
+  artist: string | null;
+  songs: string[] | null;
+};
+```
+
+This is almost there, we just need to make each property optional. We can do this by adding a `?` after the key:
```

Comment:
it would also be helpful to show that u can even *remove* optionality with `-?` (and that you can also do a similar thing with `-readonly`)